Python/Product/Uwp/Project/PythonUwpProject.cs

```diff
@@ -64,9 +67,33 @@ partial class PythonUwpProject :
         /// information from the project)
         /// </summary>
         protected override void InitializeForOuter(string fileName, string location, string name, uint flags, ref Guid guidProject, out bool cancel) {
+            var pythonProject = this.GetProject().GetPythonProject();
+            var msbuildProject = pythonProject.GetMSBuildProjectInstance();
+            msbuildProject.Build("CreatePythonUwpIoTPythonEnv", null);
+
+            var sitePackagesDir = Path.Combine(pythonProject.ProjectDirectory, PythonUwpConstants.InterpreterRelativePath, PythonUwpConstants.InterpreterLibPath, "site-packages");
+            var sitePackageDirInfo = new DirectoryInfo(sitePackagesDir);
+            if (sitePackageDirInfo.Exists) {
+                sitePackageWatcher = new FileSystemWatcher {
+                    IncludeSubdirectories = true,
+                    Path = sitePackagesDir,
+                };
+
+                sitePackageWatcher.Created += SitePackageWatcher_Changed;
+                sitePackageWatcher.Changed += SitePackageWatcher_Changed;
+                sitePackageWatcher.Deleted += SitePackageWatcher_Changed;
+                sitePackageWatcher.Renamed += SitePackageWatcher_Changed;
+                sitePackageWatcher.EnableRaisingEvents = true;
+            }
+
             base.InitializeForOuter(fileName, location, name, flags, ref guidProject, out cancel);
         }
 
+        private void SitePackageWatcher_Changed(object sender, System.IO.FileSystemEventArgs e) {
+            var bps = this._innerProject as IVsBuildPropertyStorage;
```

Comment:
Practically everything that goes through an `IVs*` interface is going to assume that it's running on the UI thread. It's very likely that any dictionary accesses behind this aren't synchronized, which means you'll get random NullReferenceExceptions (on a background thread, which will instantly terminate VS).

Unless you can step through this and show that it's always hitting a code path that is synchronized. In that case, you should be able to use a hard cast since the incoming object must always support the interface.